jstz_api/src/text_encoder.rs

```diff
@@ -1,71 +1,249 @@
-        let encoded = DEFAULT_ENGINE.encode(&str);
-        Ok(encoded.into())
+
+    fn init(class: &mut ClassBuilder<'_, '_>) -> JsResult<()> {
+        let read = accessor!(
+            class.context(),
+            TextEncoderEncodeIntoResult,
+            "read",
+            get:((x, _context) => Ok(JsBigInt::new(x.read).into()))
+        );
+        let written = accessor!(
+            class.context(),
+            TextEncoderEncodeIntoResult,
+            "written",
+            get:((x, _context) => Ok(JsBigInt::new(x.written).into()))
+        );
+        class
+            .accessor(js_string!("read"), read, Attribute::all())
+            .accessor(js_string!("written"), written, Attribute::all());
+
+        Ok(())
     }
 }
 
+// Note: Using Box as a way to get a fixed size array, probably wrong
+type Uint8Array = Box<[u8]>;
+
+impl TextEncoder {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message(
+                        "Failed to convert js value into rust type `TextEncoder`",
+                    )
+                    .into()
+            })
+    }
+
+    fn encoding() -> &'static str {
+        "utf-8"
+    }
+
+    fn encode(input: Option<String>) -> Uint8Array {
+        let s = input.unwrap_or_else(|| "".to_string());
+        s.into_bytes().into_boxed_slice()
+    }
+
+    fn encode_into(
+        source: String,
+        destination: &mut Uint8Array,
+    ) -> TextEncoderEncodeIntoResult {
+        let source = source.into_bytes().into_boxed_slice();
+        let mut count = 0u128;
+        for (i, dst_elt) in destination.iter_mut().enumerate() {
+            if let Some(src_elt) = source.get(i) {
+                count += 1;
+                *dst_elt = *src_elt;
+            }
+        }
+        TextEncoderEncodeIntoResult {
+            read: count,
+            written: count,
+        }
+    }
+}
+#[derive(Default, Clone, Trace, Finalize)]
 pub struct TextEncoderApi;
 impl TextEncoderApi {
-    const NAME: &str = "TextEncoder";
-    fn atob(_: &JsValue, args: &[JsValue], _: &mut Context) -> JsResult<JsValue> {
-        let data = args
-            .get_or_undefined(0)
-            .as_string()
-            .ok_or_else(|| JsNativeError::typ().with_message("expected string"))?;
-        let result = TextEncoder::atob(data)?;
-        Ok(result.into())
+    fn encode(_: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        let arg = match args.get(0) {
+            None => None,
+            Some(arg) => {
+                let arg = arg
+                    .as_string()
+                    .ok_or_else(|| JsNativeError::typ().with_message("expected string"))?
+                    .to_std_string_escaped();
+                Some(arg)
+            }
+        };
```

Comment:
This should be supported by `args.get_or_undefined(0).try_js_into(context)`

Reply:
done